### ISSUE TITLE
CUDA version mismatch

### ISSUE THREAD

**Issue by awsaf49** [SHI-Labs/NATTEN#31]
I am trying to install `natten` in kaggle for `OneFormer`. Kaggle has cuda `12.1` version.
code:
```shell
!pip3 install natten -f https://shi-labs.com/natten/wheels/cu121/torch2.0.0/index.html
```
error:
```shell
Looking in links: https://shi-labs.com/natten/wheels/cu121/torch2.0.0/index.html
Collecting natten
  Downloading natten-0.14.6.tar.gz (505 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 505.7/505.7 kB 9.1 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
Requirement already satisfied: packaging in /opt/conda/lib/python3.10/site-packages (from natten) (23.1)
Building wheels for collected packages: natten
  Building wheel for natten (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [70 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-3.10
      creating build/lib.linux-x86_64-3.10/natten
      copying src/natten/__init__.py -> build/lib.linux-x86_64-3.10/natten
      copying src/natten/natten1d.py -> build/lib.linux-x86_64-3.10/natten
      copying src/natten/natten2d.py -> build/lib.linux-x86_64-3.10/natten
      copying src/natten/functional.py -> build/lib.linux-x86_64-3.10/natten
      copying src/natten/flops.py -> build/lib.linux-x86_64-3.10/natten
      creating build/lib.linux-x86_64-3.10/natten/csrc
      copying src/natten/csrc/natten2dav.h -> build/lib.linux-x86_64-3.10/natten/csrc
      copying src/natten/csrc/natten2dqkrpb.h -> build/lib.linux-x86_64-3.10/natten/csrc
      copying src/natten/csrc/context.h -> build/lib.linux-x86_64-3.10/natten/csrc
      copying src/natten/csrc/natten1dav.h -> build/lib.linux-x86_64-3.10/natten/csrc
      copying src/natten/csrc/natten.cpp -> build/lib.linux-x86_64-3.10/natten/csrc
      copying src/natten/csrc/natten1dqkrpb.h -> build/lib.linux-x86_64-3.10/natten/csrc
      creating build/lib.linux-x86_64-3.10/natten/csrc/cpu
      copying src/natten/csrc/cpu/natten_cpu_commons.h -> build/lib.linux-x86_64-3.10/natten/csrc/cpu
      copying src/natten/csrc/cpu/natten2dav_cpu_kernel.cpp -> build/lib.linux-x86_64-3.10/natten/csrc/cpu
      copying src/natten/csrc/cpu/natten1dav_cpu_kernel.cpp -> build/lib.linux-x86_64-3.10/natten/csrc/cpu
      copying src/natten/csrc/cpu/natten2dqkrpb_cpu_kernel.cpp -> build/lib.linux-x86_64-3.10/natten/csrc/cpu
      copying src/natten/csrc/cpu/natten1dqkrpb_cpu_kernel.cpp -> build/lib.linux-x86_64-3.10/natten/csrc/cpu
      creating build/lib.linux-x86_64-3.10/natten/csrc/cuda
      copying src/natten/csrc/cuda/natten1dqkrpb_cuda_kernel.cu -> build/lib.linux-x86_64-3.10/natten/csrc/cuda
      copying src/natten/csrc/cuda/natten1dav_cuda_kernel.cu -> build/lib.linux-x86_64-3.10/natten/csrc/cuda
      copying src/natten/csrc/cuda/natten_commons.cuh -> build/lib.linux-x86_64-3.10/natten/csrc/cuda
      copying src/natten/csrc/cuda/natten2dqkrpb_cuda_kernel.cu -> build/lib.linux-x86_64-3.10/natten/csrc/cuda
      copying src/natten/csrc/cuda/natten2dav_cuda_kernel.cu -> build/lib.linux-x86_64-3.10/natten/csrc/cuda
      running build_ext
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-rowagv61/natten_17f593ad740d4ad38ccb35df0695cbe1/setup.py", line 118, in <module>
          setup(
        File "/opt/conda/lib/python3.10/site-packages/setuptools/__init__.py", line 153, in setup
          return distutils.core.setup(**attrs)
        File "/opt/conda/lib/python3.10/distutils/core.py", line 148, in setup
          dist.run_commands()
        File "/opt/conda/lib/python3.10/distutils/dist.py", line 966, in run_commands
          self.run_command(cmd)
        File "/opt/conda/lib/python3.10/distutils/dist.py", line 985, in run_command
          cmd_obj.run()
        File "/opt/conda/lib/python3.10/site-packages/wheel/bdist_wheel.py", line 343, in run
          self.run_command("build")
        File "/opt/conda/lib/python3.10/distutils/cmd.py", line 313, in run_command
          self.distribution.run_command(command)
        File "/opt/conda/lib/python3.10/distutils/dist.py", line 985, in run_command
          cmd_obj.run()
        File "/opt/conda/lib/python3.10/distutils/command/build.py", line 135, in run
          self.run_command(cmd_name)
        File "/opt/conda/lib/python3.10/distutils/cmd.py", line 313, in run_command
          self.distribution.run_command(command)
        File "/opt/conda/lib/python3.10/distutils/dist.py", line 985, in run_command
          cmd_obj.run()
        File "/opt/conda/lib/python3.10/site-packages/setuptools/command/build_ext.py", line 79, in run
          _build_ext.run(self)
        File "/opt/conda/lib/python3.10/site-packages/Cython/Distutils/old_build_ext.py", line 186, in run
          _build_ext.build_ext.run(self)
        File "/opt/conda/lib/python3.10/distutils/command/build_ext.py", line 340, in run
          self.build_extensions()
        File "/opt/conda/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 499, in build_extensions
          _check_cuda_version(compiler_name, compiler_version)
        File "/opt/conda/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 387, in _check_cuda_version
          raise RuntimeError(CUDA_MISMATCH_MESSAGE.format(cuda_str_version, torch.version.cuda))
      RuntimeError:
      The detected CUDA version (12.1) mismatches the version that was used to compile
      PyTorch (11.3). Please make sure to use the same CUDA versions.
```

**Comment by alihassanijr** [comment]
Hello and thank you for your interest.

Unfortunately we only support CUDA versions that PyTorch's stable builds were compiled with.
Since there are so many possible combinations, PyTorch typically picks 2 or 3 CUDA versions at maximum and releases binaries with those, therefore, we do the same (except we pick exactly the versions torch picked.)
As far as I'm aware PyTorch still has no stable build with CUDA 12.X.

That said, we have tested CUDA 12.0 in the past by building pytorch from source, but again the issue with releasing binaries is stability.

The error here is also from PyTorch, not from NATTEN. The wheel URL is ignored because it doesn't exist, and when it tries to compile NATTEN from source (requires PyTorch's C++/CUDA backend), it fails because your installed pytorch was compiled with CUDA 11.3, and your CUDA compiler is 12.1.

### Solution IF PyTorch already works
Please find your PyTorch version:
```python3
import torch
print(torch.__version__)
```

and install the NATTEN build associated with it. The error indicates the CUDA version it was compiled with is 11.3, so it's likely that you're either on torch 1.10, 1.11, 1.12 or 1.12.1.
Once you're sure, you can refer to our website and get a pip install command with the correct wheel URL: [shi-labs.com/natten](https://shi-labs.com/natten/)

OR alternatively do:
```
pip3 install natten -f https://shi-labs.com/natten/wheels/cu113/torch1.12.1/index.html 
```
and replace the torch version.

**Comment by awsaf49** [comment]
It seems It's cuda 12> issue. In kaggle installing `natten` with older version of cuda gives error. So I guess there is no fix for now :(

**Comment by alihassanijr** [comment]
No, just to be clear, the issue is not CUDA. It's that the version of CUDA you're on doesn't match your PyTorch version (and PyTorch has no stable releases with CUDA 12 or 12.1 binaries).
You can try the solution I wrote above and install NATTEN with precompiled binaries that match your PyTorch install, but it is possible for you to run into issues with CUDA in general because of the mismatch.

Another solution you might be able to try is setting up a conda environment with an older CUDA version (11.7 or 11.8).

**Comment by alihassanijr** [comment]
Sorry if it was confusing, but I didn't suggest installing NATTEN for Torch 2.0. You should always install NATTEN for the --exact-- torch version you have installed.

The original issue you posted was caused because the URL was incorrect; as in there's no PyTorch build compiled with CUDA 12.1, hence there is no NATTEN build compiled with CUDA 12.1. Because of that, pip tried to compile NATTEN on your system, which has CUDA 12.1, and because of the mismatch with PyTorch's binaries, you get the mismatch error.

My suggestion was to check your torch version (because you obviously have torch installed), and the cuda version associated with it, and install the corresponding NATTEN.

Based on your original post, you should be on torch 1.10, 1.11, 1.12 or 1.12.1. Not 2.0.

Regarding Kaggle's environment, I'm not sure what can be done about that unfortunately. If it doesn't allow conda environments there's very little that can be done about the local CUDA version. But again, NATTEN should work as long as you're on a stable and working PyTorch install with CUDA. All you have to do is check your torch version as I mentioned, and refer to [our website](https://shi-labs.com/natten/) and get the correctly matching NATTEN build.

**Comment by awsaf49** [comment]
Sorry, perhaps I couldn't make it clear. 
In kaggle I am using PyTorch `2.0.0` thus I have installed `natten` for corresponding pytorch==2.0.0
You can check out this [notebook](https://www.kaggle.com/code/awsaf49/oneformer-issue) to reproduce the error,

**Comment by alihassanijr** [comment]
The issue in your notebook seems to be that you're on a non-pip build of PyTorch (probably conda). The issue with those is that they create such problems when there's a mismatch between your CUDA compiler version and the version of torch you install, specifically when torch doesn't yet support your CUDA version.

If you look at your torch version:
```
import torch
torch.__version__
```
It prints out just `2.0.0` -- not `2.0.0+cuXXX`. This indicates that you probably didn't install torch from pip, and I'm not too sure about what the best way to check which cuda driver you have installed is (been a while since I used notebooks, and I haven't used Kaggle's environment before).

Anyway, the fix is simple.
Just uninstall your torch, reinstall via pip, and then install the corresponding NATTEN.

For example, I did:
```
!pip uninstall torch torchvision torchaudio -y
!pip install torch==2.0.0+cu118 torchvision==0.15.1+cu118 torchaudio==2.0.1 --index-url https://download.pytorch.org/whl/cu118
```
(source: [PyTorch.org](https://pytorch.org/get-started/locally/))

Had to restart the notebook to make sure the torch import wasn't cached.
Then:
```
!pip3 install natten -f https://shi-labs.com/natten/wheels/cu118/torch2.0.0/index.html 
```
(source: [our website](https://shi-labs.com/natten/))

And it worked just fine.
```python3
from natten import NeighborhoodAttention2D

na2d = NeighborhoodAttention2D(dim=128, num_heads=4, kernel_size=7, dilation=1).cuda()
x = torch.randn((1, 16, 16, 128)).cuda()

y = na2d(x)

out = y.sum()

print(y.shape)
```
```
torch.Size([1, 16, 16, 128])
```
